README.md
# NAME

Catmandu::Importer::getJSON - load JSON-encoded data from a server using a GET HTTP request

# STATUS

[![Build Status](https://travis-ci.org/nichtich/Catmandu-Importer-getJSON.png)](https://travis-ci.org/nichtich/Catmandu-Importer-getJSON)
[![Coverage Status](https://coveralls.io/repos/nichtich/Catmandu-Importer-getJSON/badge.png)](https://coveralls.io/r/nichtich/Catmandu-Importer-getJSON)
[![Kwalitee Score](http://cpants.cpanauthors.org/dist/Catmandu-Importer-getJSON.png)](http://cpants.cpanauthors.org/dist/Catmandu-Importer-getJSON)

# SYNOPSIS

The following three examples are equivalent:

    Catmandu::Importer::getJSON->new(
        file => \"http://example.org/alice.json\nhttp://example.org/bob.json"
    )->each(sub { my ($record) = @_; ... );

    Catmandu::Importer::getJSON->new(
        url  => "http://example.org",
        file => \"/alice.json\n/bob.json"
    )->each(sub { my ($record) = @_; ... );

    Catmandu::Importer::getJSON->new(
        url  => "http://example.org/{name}.json",
        file => \"{\"name\":\"alice\"}\n{\"name\":\"bob\"}"
    )->each(sub { my ($record) = @_; ... );

For more convenience the [catmandu](https://metacpan.org/pod/catmandu) command line client can be used:

    echo http://example.org/alice.json | catmandu convert getJSON to YAML
    catmandu convert getJSON --from http://example.org/alice.json to YAML
    catmandu convert getJSON --dry 1 --url http://{domain}/robots.txt < domains

# DESCRIPTION

This [Catmandu::Importer](https://metacpan.org/pod/Catmandu::Importer) performs a HTTP GET request to load JSON-encoded
data from a server. The importer expects a line-separated input. Each line
corresponds to a HTTP request that is mapped to a JSON-record on success. The
following input formats are accepted:

- plain URL

    A line that starts with "`http://`" or "`https://`" is used as plain URL.

- URL path

    A line that starts with "`/`" is appended to the configured **url** parameter.

- variables

    A JSON object with variables to be used with an URL template or as HTTP query
    parameters. For instance the input line `{"name":"Karl Marx"}` with URL
    `http://api.lobid.org/person` or the input line
    `{"entity":"person","name":"Karl Marx"}` with URL template
    `http://api.lobid.org/{entity}{?id}{?name}{?q}` are both expanded to
    [http://api.lobid.org/person?name=Karl+Marx](http://api.lobid.org/person?name=Karl+Marx).

If the JSON data returned in a HTTP response is a JSON array, its elements are
imported as multiple items. If a JSON object is returned, it is imported as one
item.

# CONFIGURATION

- url

    An [URI](https://metacpan.org/pod/URI) or an URI templates ([URI::Template](https://metacpan.org/pod/URI::Template)) as defined by
    [RFC 6570](http://tools.ietf.org/html/rfc6570) to load JSON from. If no **url**
    is configured, plain URLs must be provided as input or option `from` must be
    used instead.

- from

    A plain URL to load JSON without reading any input lines.

- timeout / agent / proxy / headers

    Optional HTTP client settings.

- client

    Instance of a [Furl](https://metacpan.org/pod/Furl) HTTP client to perform requests with.

- dry

    Don't do any HTTP requests but return URLs that data would be queried from.

- file / fh

    Input to read lines from (see [Catmandu::Importer](https://metacpan.org/pod/Catmandu::Importer)). Defaults to STDIN.

- fix

    An optional fix to be applied on every item (see [Catmandu::Fix](https://metacpan.org/pod/Catmandu::Fix)).

- wait

    Number of seconds to wait between requests.

- cache

    Cache JSON response of URLs to not request the same URL twice. HTTP error
    codes in the 4xx range (e.g. 404) are also cached but 5xx errors are not.

    The value of this option can be any objects that implements method `get` and
    `set` (e.g. `CHI`), an existing directory for file caching, a true value to
    enable global in-memory-caching, or a false value to disable caching (default).

    File caching uses file names based on MD5 of an URL so for instance
    `http://example.org/` is cached as `4389382917e51695b759543fdfd5f690.json`.

- warn

    Show error messages on the standard error.

# METHODS

## time

Returns the UNIX timestamp right before the last request. This can be used for
instance to add timestamps or the measure how fast requests were responded.

## construct\_url( \[ $base\_url, \] $vars\_url\_or\_path )

Returns an URL given a hash reference with variables, a plain URL or an URL
path. The optional first argument can be used to override option `url`.

    $importer->construct_url( %query_vars )
    $importer->construct_url( $importer->url, %query_vars ) # equivalent

## request($url)

Perform a HTTP GET request of a given URL including logging, caching, request
hook etc. Returns a hash/array reference or `undef`.

# EXTENDING

This importer provides two methods to filter requests and responses,
respectively. See [Catmandu::Importer::Wikidata](https://metacpan.org/pod/Catmandu::Importer::Wikidata) for an example.

## request\_hook

Gets a whitespace-trimmed input line and is expected to return an unblessed
hash reference, an URL, or undef. Errors are catched and treated equal to
undef.

## response\_hook

Gets the queried response object and is expected to return an object.

# LOGGING

URLs are emitted before each request on DEBUG log level.

# LIMITATIONS

Future versions of this module may also support asynchronous HTTP fetching
modules such as [HTTP::Async](https://metacpan.org/pod/HTTP::Async), for retrieving multiple URLs at the same time.

# SEE ALSO

[Catmandu::Fix::get\_json](https://metacpan.org/pod/Catmandu::Fix::get_json) provides this importer as fix function.

# COPYRIGHT AND LICENSE

Copyright Jakob Voß, 2014-

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.


Changes
Changelog for Catmandu-Importer-getJSON

{{$NEXT}}

0.52  2019-08-30 11:29:50 CEST
 - Fix test issues

0.51  2019-01-23 20:12:35 CET
 - Disable default warning enabled (issue #15) 

0.50  2015-05-19 12:48:46 CEST
 - more powerful get_json fix function
 - new public methods request and construct_url 
 - proceed on URL errors when reading URLs from stream
 - made in-memory cache global

0.43  2015-05-10 22:12:01 CEST
 - added caching

0.42  2015-04-06 16:52:48 CEST
 - fix dependencies and URI::Template changes

0.41  2014-10-26 10:35:49 CET
 - migrated to Dist::Milla
 - include missing version (issue #9)

0.40 2014-05-16T09:08:44
 - document Catmandu::Fix::get_json
 - enable logging and option 'wait'
 - bump version
 - cpanfile [vpeil]
 - get_json fix added [vpeil]

0.3 2014-04-23T19:28:43
 - fix option 'from'

0.2 2014-04-22T18:08:25
 - version number
 - introduced request and response hooks
 - new option --dry
 - option --from closes issue #1
 - minor documentation changes

0.1 2014-01-21T11:08:38
 - use UTF-8
 - initial version



lib/Catmandu/Fix/get_json.pm
package Catmandu::Fix::get_json;

our $VERSION = '0.52';

use Catmandu::Sane;
use Catmandu::Util;
use Moo;
use Catmandu::Fix::Has;
use Catmandu::Importer::getJSON;

with "Catmandu::Fix::Base";

has url => ( fix_arg => 1, default => sub { "url" } );

has dry     => ( fix_opt => 1 );
has cache   => ( fix_opt => 1 );
has timeout => ( fix_opt => 1, default => sub { 10 } );
has agent   => ( fix_opt => 1 );
has proxy   => ( fix_opt => 1 );
has wait    => ( fix_opt => 1 );

has vars => ( fix_opt => 1 );
has path => ( fix_opt => 1 );

has importer => ( is => 'ro', lazy => 1, builder => 1 );

sub _build_importer {
    my ($self) = @_;

    my %options = (
        dry     => $self->dry,
        cache   => $self->cache,
        timeout => $self->timeout,
        agent   => $self->agent,
        proxy   => $self->proxy,
        wait    => $self->wait,
    );

    # URL template or plain URL
    if ( $self->url =~ qr{^https?://} ) {
        $options{ $self->vars ? 'url' : 'from' } = $self->url;
    }
    Catmandu::Importer::getJSON->new(%options);
}

sub BUILD {
    my ($self) = @_;
    unless ( defined $self->path ) {
        $self->{path} = $self->url =~ qr{^https?://} ? '' : $self->url;
    }
}

sub emit {
    my ( $self, $fixer ) = @_;
    my $path     = $fixer->split_path( $self->path );
    my $importer = $fixer->capture( $self->importer );

    # plain URL
    if ( $self->importer->from ) {
        return $fixer->emit_create_path(
            $fixer->var,
            $path,
            sub {
                sprintf '%s = %s->request(%s->from) // { };',
                  shift, $importer, $importer;
            }
        );
    }

    # URL template or base URL
    if ( $self->importer->url ) {
        my $tpl = $fixer->split_path( $self->vars );
        my $url = $fixer->generate_var;

        return $fixer->emit_create_path(
            $fixer->var,
            $tpl,
            sub {
                my $tpl = shift;
                "my $url;"
                  . "if (is_hash_ref($tpl) or is_string($tpl) and $tpl !~ qr{^https?://}) {"
                  . "  $url = ${importer}->construct_url($tpl) " . "}"
                  . $fixer->emit_create_path(
                    $fixer->var,
                    $path,
                    sub {
                        sprintf '%s = %s ? %s->request(%s) // {} : {};',
                          shift, $url, $importer, $url;
                    }
                  );
            }
        );
    }

    # URL from field
    my $url = $fixer->split_path( $self->url );

    return $fixer->emit_create_path(
        $fixer->var,
        $url,
        sub {
            if ( $self->vars ) {
                my $base = shift;
                my $tpl  = $fixer->split_path( $self->vars );
                my $url  = $fixer->generate_var;

                return $fixer->emit_create_path(
                    $fixer->var,
                    $tpl,
                    sub {
                        my $tpl = shift;
                        "my $url;"
                          . "if (is_hash_ref($tpl) or is_string($tpl) and $tpl !~ qr{^https?://}) {"
                          . "  $url = ${importer}->construct_url($base, $tpl); "
                          . "}"
                          . $fixer->emit_create_path(
                            $fixer->var,
                            $path,
                            sub {
                                sprintf '%s = %s ? %s->request(%s) // {} : {};',
                                  shift, $url, $importer, $url;
                            }
                          );
                    }
                );
            }
            else {
                my $url = shift;
                return $fixer->emit_create_path(
                    $fixer->var,
                    $path,
                    sub {
                        sprintf '%s = %s->request(%s) // { };',
                          shift, $importer, $url;
                    }
                );
            }
        }
    );

}

1;
__END__

=head1 NAME

Catmandu::Fix::get_json - get JSON data from an URL as fix function

=head1 SYNOPSIS

	# fetches a hash or array
	get_json("http://example.com/json")

	# stores it in path.key
	get_json("http://example.com/json", path: path.key)

    # add URL query parameters or URL path from config
	get_json("http://example.com/", vars: config)

    # fill URL template fields from config
	get_json("http://example.com/{name}.json", vars: config)

  	# get URL or URL template from a field
	get_json(field.name)
 
=head1 DESCRIPTION

This L<Catmandu::Fix> provides a method to fetch JSON data from an URL. The
response is added as new item or to a field of the current item.

=head1 OPTIONS

The first argument must be an URL, an URL template, or a field where to take
an URL or URL template from. Additional options include:

=over

=item path

Field to store result JSON in. If not given or set to the empty string, the
whole item is replaced by the fetched JSON response. If the first argument is a
field, the same field is used as C<path> by default.

=item vars

Field to get URL template variables, URL query parameters or an URL path
expression from. This option is required if the first argument is an URL
template. 

=back

The fix function also supports options C<dry>, C<cache>, C<timeout>, C<agent>,
C<proxy>, and C<wait> as documented in L<Catmandu::Importer::getJSON>.  Options
C<client>, C<headers>, and C<warn> are not supported.

=cut


lib/Catmandu/Importer/HTTP.pm
package Catmandu::Importer::HTTP;
#ABSTRACT: import items via HTTP GET requests
#VERSION

use Catmandu::Sane;
use Moo;
use JSON;
use Furl;
use Scalar::Util qw(blessed);
use URI::Template;

with 'Catmandu::Importer';

has url     => ( is => 'rw', trigger => 1 );
has from    => ( is => 'ro');
has timeout => ( is => 'ro', default => sub { 10 } );
has agent   => ( is => 'ro' );
has proxy   => ( is => 'ro' );
has dry     => ( is => 'ro' );
has headers => ( is => 'ro' );
has wait    => ( is => 'ro' );
has client  => (
    is => 'ro',
    lazy => 1,
    builder => sub { 
        Furl->new( 
            map { $_ => $_[0]->{$_} } grep { defined $_[0]->{$_} }
            qw(timeout agent proxy),
        ) 
    }
);

has json => (is => 'ro', default => sub { JSON->new->utf8(1) } );

has time => (is => 'rw');

sub _trigger_url {
    my ($self, $url) = @_;

    if (!blessed $url) {
        $url = URI::Template->new($url) unless blessed $url;
    }

    if ($url->isa('URI::Template')) {
        unless ($url->variables) {
            $url = URI->new("$url");
        }
    }

    $self->{url} = $url;
}

sub generator {
    my ($self) = @_;
    
    if ($self->from) {
        return sub {
            state $data = do {
                my $r = $self->_query_url($self->from);
                (ref $r // '') eq 'ARRAY' ? $r : [$r];
            };
            return shift @$data;
        }
    }

    sub {
        state $fh = $self->fh;
        state $data;

        if ( $data and ref $data eq 'ARRAY' and @$data ) {
            return shift @$data;
        }

        my $line = <$fh> // return;
        my $url = $self->_construct_url($line) // return;

        $data = $self->_query_url($url);

        return (ref $data // '') eq 'ARRAY' ? shift @$data : $data;
    }
}

sub request_hook {
    my ($self, $line) = @_;

    if ($line =~ /^\s*{/) {
        return $self->json->decode($line); 
    } else {
        my $url;

        # plain URL
        if ( $line =~ /^https?:\// ) {
            $url = URI->new($line);
        # URL path (and optional query)
        } elsif ( $line =~ /^\// ) {
            $url = "".$self->url;
            $url =~ s{/$}{}; 
            $line =~ s{\s+$}{};
            $url = URI->new($url . $line);
        }
 
        return $url;
    }
}

sub _construct_url {
    my ($self, $line) = @_;

    chomp $line;
    $line =~ s/^\s+|\s+$//g;

    my $request = $self->request_hook($line);
    my $url;

    # Template or query variables
    if (ref $request and not blessed($request)) {
        $url = $self->url;
        if ($url->isa('URI::Template')) {
            $url = $url->process($request);
        } else {
            $url = $url->clone;
            $url->query_form($request);
        }         
    } else {
        $url = $request;
    }
  
    warn "failed to _construct URL from: '$line'\n" unless $url;

    return $url;
}


sub _query_url {
    my ($self, $url) = @_;

    $self->log->debug($url);

    if ( $self->dry ) {
        return { url => "$url" };
    }

    if ( $self->wait and $self->time ) {
        my $elapsed = ($self->time // time) - time;
        sleep( $self->wait - $elapsed );
    }
    $self->time(time);

    my $response = $self->client->get($url, $self->headers // []);
    unless ($response->is_success) {
        warn "request failed: $url\n";
        return;
    }

    my $data = { 
        code     => $response->code,
        message  => $response->message,
        protocol => $response->protocol,
        headers  => [$response->headers->flatten],
        content  => $response->decoded_content,
    };

    $self->response_hook($data);
}

sub response_hook { $_[1] }

sub parse {
    my ($self, $response) = @_;
    $response;
}

1;

=head1 DESCRIPTION

This L<Catmandu::Importer> performs a HTTP GET request to load items from a
server. See L<Catmandu::Importer::getJSON> and L<Catmandu::Importer::getXML>
for specialized variants that load JSON- or XML-encoded items, respectively. 

The importer expects a line-separated input. Each line corresponds to a HTTP
request that is mapped to a JSON-record on success. The following input formats
are accepted:

=over

=item plain URL

A line that starts with "C<http://>" or "C<https://>" is used as plain URL.

=item URL path

A line that starts with "C</>" is appended to the configured B<url> parameter.

=item variables

A JSON object with variables to be used with an URL template or as HTTP query
parameters. For instance the input line C<< {"name":"Karl Marx"} >> with URL
C<http://api.lobid.org/person> or the input line 
C<< {"entity":"person","name":"Karl Marx"} >> with URL template
C<http://api.lobid.org/{entity}{?id}{?name}{?q}> are both expanded to
L<http://api.lobid.org/person?name=Karl+Marx>.

=back

=head1 CONFIGURATION

=over

=item url

An L<URI> or an URI templates (L<URI::Template>) as defined by 
L<RFC 6570|http://tools.ietf.org/html/rfc6570> to load JSON from. If no B<url>
is configured, plain URLs must be provided as input or option C<from> must be
used instead.

=item from

A plain URL to load from without reading any input lines.

=item timeout / agent / proxy / headers

Optional HTTP client settings.

=item client

Instance of a L<Furl> HTTP client to perform requests with.

=item dry

Don't do any HTTP requests but return URLs that data would be queried from. 

=item file / fh

Input to read lines from (see L<Catmandu::Importer>). Defaults to STDIN.

=item fix

An optional fix to be applied on every item (see L<Catmandu::Fix>).

=item wait

Number of seconds to wait between requests.

=back

=head1 METHODS

=head2 time

Returns the UNIX timestamp right before the last request. This can be used for
instance to add timestamps or the measure how fast requests were responded.

=head1 EXTENDING

This importer provides two methods to filter requests and responses,
respectively. See L<Catmandu::Importer::Wikidata> for an example.

=head2 request_hook

Gets a whitespace-trimmed input line and is expected to return an unblessed
object or an URL.

=head2 response_hook

Gets the HTTP response in form of a HASH reference as following to transform to
an item:

    {
        protocol => 'HTTP/1.1',
        code     => 200,
        message  => 'Ok',
        headers  => [ key => 'value', key => 'value', ... ],
        content  => '...decoded response content...'
    }

Subclasses should implement their own variant of this method. See
L<Catmandu::Importer::getJSON> and L<Catmandu::Importer::getJSON> for simple
examples.

=head1 LOGGING

URLs are emitted before each request on DEBUG log level.

=head1 LIMITATIONS

Error handling is very limited.

Future versions of this module may also support asynchronous HTTP fetching
modules such as L<HTTP::Async>, for retrieving multiple URLs at the same time.

=encoding utf8


lib/Catmandu/Importer/getXML.pm
package Catmandu::Importer::getXML;
#ABSTRACT: import XML via HTTP GET request
#VERSION

use Moo;
use XML::Struct::Reader;
#use Catmandu::Importer::XML;
#TODO: Catmandu::Importer::Foo->new(%args, file => \$content)->to_array;
# getCSV, getXLS...

extends 'Catmandu::Importer::HTTP';

has '+headers' => (
    is => 'ro',
    default => sub { [ 'Accept' => 'application/xml' ] }
);

has simple => (is => 'ro', default => sub { 1 });

sub response_hook {
    my ($self, $response) = @_;

    my $reader = XML::Struct::Reader->new(
        simple => $self->simple,
        from => \$response->{content},
    );

    $reader->read;
}

1;

=head1 SYNOPSIS

    echo http://example.org/data.xml | catmandu convert getXML to YAML
    catmandu convert getXML --from http://example.org/data.xml to YAML

=head1 DESCRIPTION

This L<Catmandu::Importer> is a L<Catmandu::Importer::HTTP> that reads XML
data with L<XML::Struct>.

=head1 CONFIGURATION

=over

=item simple

Parse XML into simple format, not preserving element order (default).

=back

=head1 SEE ALSO

L<Catmandu::XML>

=encoding utf8


lib/Catmandu/Importer/getJSON.pm
package Catmandu::Importer::getJSON;

our $VERSION = '0.51';
our $CACHE;

use Catmandu::Sane;
use Moo;
use JSON;
use Furl;
use Scalar::Util qw(blessed);
use URI::Template;

with 'Catmandu::Importer';

has url => (
    is      => 'rw',
    trigger => sub {
        $_[0]->{url} = _url_template_or_url( $_[1] );
    }
);

has from    => ( is => 'ro' );
has timeout => ( is => 'ro', default => sub { 10 } );
has agent   => ( is => 'ro' );
has proxy   => ( is => 'ro' );
has dry     => ( is => 'ro' );
has headers => (
    is      => 'ro',
    default => sub { [ 'Accept' => 'application/json' ] }
);
has wait => ( is => 'ro' );
has cache => ( is => 'ro', trigger => 1 );
has client => (
    is      => 'ro',
    lazy    => 1,
    builder => sub {
        Furl->new(
            map { $_ => $_[0]->{$_} } grep { defined $_[0]->{$_} }
              qw(timeout agent proxy),
        );
    }
);
has json => ( is => 'ro', default => sub { JSON->new->utf8(1) } );
has time => ( is => 'rw' );
has warn => ( is => 'ro', default => sub { 0 } );

sub _url_template_or_url {
    my ($url) = @_;

    if ( !blessed $url) {
        $url = URI::Template->new($url);
    }

    if ( $url->isa('URI::Template') ) {
        unless ( my @variables = $url->variables ) {
            $url = URI->new("$url");
        }
    }
    return $url;
}

{

    package Importer::getJSON::MemoryCache;
    use JSON;
    our $JSON = JSON->new->utf8;
    sub new { bless {}, $_[0] }

    sub get {
        eval { $JSON->decode( $_[0]->{ $_[1] } ) };
    }
    sub set { $_[0]->{ $_[1] } = ref $_[2] ? $JSON->encode( $_[2] ) : '' }
}
$CACHE = Importer::getJSON::MemoryCache->new;

{

    package Importer::getJSON::FileCache;
    use JSON;
    use Catmandu::Util qw(read_json);
    use Digest::MD5 qw(md5_hex);
    our $JSON = JSON->new->utf8;

    sub new {
        my ( $class, $dir ) = @_;
        $dir =~ s{/$}{};
        bless { dir => $dir }, $class;
    }

    sub file {
        my ( $self, $url ) = @_;
        $self->{dir} . '/' . md5_hex($url->as_string) . '.json';
    }

    sub get {
        eval { read_json( $_[0]->file( $_[1] ) ) };
    }

    sub set {
        my ( $self, $url, $data ) = @_;
        open my $fh, ">", $self->file($url);
        print $fh ( ref $data ? $JSON->encode($data) : '' );
    }
}

sub _trigger_cache {
    my ( $self, $cache ) = @_;

    if ( blessed $cache and $cache->can('get') and $cache->can('set') ) {

        # use cache object
    }
    elsif ( $cache and -d $cache ) {
        $cache = Importer::getJSON::FileCache->new($cache);
    }
    elsif ($cache) {
        $cache = $CACHE;
    }

    $self->{cache} = $cache;
}

sub generator {
    my ($self) = @_;

    if ( $self->from ) {
        return sub {
            state $data = do {
                my $r = $self->request( $self->from );
                ( ref $r // '' ) eq 'ARRAY' ? $r : [$r];
            };
            return shift @$data;
          }
    }

    sub {
        state $fh = $self->fh;
        state $data;

        if ( $data and ref $data eq 'ARRAY' and @$data ) {
            return shift @$data;
        }

        my $url;
        until ($url) {
            my $line = <$fh> // return;
            chomp $line;
            $line =~ s/^\s+|\s+$//g;
            next if $line eq '';    # ignore empty lines

            my $request = eval { $self->request_hook($line) };
            $url = $self->construct_url($request);
            if ( !$url ) {
                warn "failed to construct URL: $line\n" if $self->warn;
                $self->log->warn("failed to construct URL: $line");
            }
        }

        $data = $self->request($url);

        return ( ref $data // '' ) eq 'ARRAY' ? shift @$data : $data;
      }
}

sub request_hook {
    my ( $self, $line ) = @_;
    return $line =~ /^\s*{/ ? $self->json->decode($line) : $line;
}

sub construct_url {
    my $self    = shift;
    my $url     = @_ > 1 ? _url_template_or_url(shift) : $self->url;
    my $request = shift;

    # Template or query variables
    if ( ref $request and not blessed $request) {
        return unless blessed $url;
        if ( $url->isa('URI::Template') ) {
            $url = $url->process($request);
        }
        else {
            $url = $url->clone;
            $url->query_form($request);
        }
        return $url;
    }
    elsif ( blessed $request and $request->isa('URI::URL') ) {
        return $request;
    }
    elsif ( $request =~ /^https?:\/\// ) {    # plain URL
        return URI->new($request);
    }
    elsif ( $request =~ /^\// ) {             # URL path (and optional query)
        $url = "$url";
        $url =~ s{/$}{};
        $request =~ s{\s+$}{};
        return URI->new( $url . $request );
    }

    return;
}

sub request {
    my ( $self, $url ) = @_;

    $self->log->debug($url);

    my $json = '';

    if ( $self->dry ) {
        return { url => "$url" };
    }

    if ( $self->cache ) {
        $json = $self->cache->get($url);
        if ( defined $json ) {
            return ref $json ? $json : undef;
        }
    }

    if ( $self->wait and $self->time ) {
        my $elapsed = ( $self->time // time ) - time;
        sleep( $self->wait - $elapsed );
    }
    $self->time(time);

    my $response = $self->client->get( $url, $self->headers );
    if ( $response->is_success ) {
        my $content = $response->decoded_content;
        my $data    = $self->json->decode($content);
        $json = $self->response_hook($data);
    }
    else {
        warn "request failed: $url\n" if $self->warn;
        $self->log->warn("request failed: $url");
        if ( $response->status =~ /^4/ ) {
            $json = '';
        }
        else {
            return;
        }
    }

    if ( $self->cache ) {
        $self->cache->set( $url, $json );
    }

    return ref $json ? $json : undef;
}

sub response_hook { $_[1] }

1;
__END__

=head1 NAME

Catmandu::Importer::getJSON - load JSON-encoded data from a server using a GET HTTP request

=begin markdown

# STATUS

[![Build Status](https://travis-ci.org/nichtich/Catmandu-Importer-getJSON.png)](https://travis-ci.org/nichtich/Catmandu-Importer-getJSON)
[![Coverage Status](https://coveralls.io/repos/nichtich/Catmandu-Importer-getJSON/badge.png)](https://coveralls.io/r/nichtich/Catmandu-Importer-getJSON)
[![Kwalitee Score](http://cpants.cpanauthors.org/dist/Catmandu-Importer-getJSON.png)](http://cpants.cpanauthors.org/dist/Catmandu-Importer-getJSON)

=end markdown

=head1 SYNOPSIS

The following three examples are equivalent:

    Catmandu::Importer::getJSON->new(
        file => \"http://example.org/alice.json\nhttp://example.org/bob.json"
    )->each(sub { my ($record) = @_; ... );

    Catmandu::Importer::getJSON->new(
        url  => "http://example.org",
        file => \"/alice.json\n/bob.json"
    )->each(sub { my ($record) = @_; ... );

    Catmandu::Importer::getJSON->new(
        url  => "http://example.org/{name}.json",
        file => \"{\"name\":\"alice\"}\n{\"name\":\"bob\"}"
    )->each(sub { my ($record) = @_; ... );

For more convenience the L<catmandu> command line client can be used:

    echo http://example.org/alice.json | catmandu convert getJSON to YAML
    catmandu convert getJSON --from http://example.org/alice.json to YAML
    catmandu convert getJSON --dry 1 --url http://{domain}/robots.txt < domains

=head1 DESCRIPTION

This L<Catmandu::Importer> performs a HTTP GET request to load JSON-encoded
data from a server. The importer expects a line-separated input. Each line
corresponds to a HTTP request that is mapped to a JSON-record on success. The
following input formats are accepted:

=over

=item plain URL

A line that starts with "C<http://>" or "C<https://>" is used as plain URL.

=item URL path

A line that starts with "C</>" is appended to the configured B<url> parameter.

=item variables

A JSON object with variables to be used with an URL template or as HTTP query
parameters. For instance the input line C<< {"name":"Karl Marx"} >> with URL
C<http://api.lobid.org/person> or the input line
C<< {"entity":"person","name":"Karl Marx"} >> with URL template
C<http://api.lobid.org/{entity}{?id}{?name}{?q}> are both expanded to
L<http://api.lobid.org/person?name=Karl+Marx>.

=back

If the JSON data returned in a HTTP response is a JSON array, its elements are
imported as multiple items. If a JSON object is returned, it is imported as one
item.

=head1 CONFIGURATION

=over

=item url

An L<URI> or an URI templates (L<URI::Template>) as defined by
L<RFC 6570|http://tools.ietf.org/html/rfc6570> to load JSON from. If no B<url>
is configured, plain URLs must be provided as input or option C<from> must be
used instead.

=item from

A plain URL to load JSON without reading any input lines.

=item timeout / agent / proxy / headers

Optional HTTP client settings.

=item client

Instance of a L<Furl> HTTP client to perform requests with.

=item dry

Don't do any HTTP requests but return URLs that data would be queried from.

=item file / fh

Input to read lines from (see L<Catmandu::Importer>). Defaults to STDIN.

=item fix

An optional fix to be applied on every item (see L<Catmandu::Fix>).

=item wait

Number of seconds to wait between requests.

=item cache

Cache JSON response of URLs to not request the same URL twice. HTTP error
codes in the 4xx range (e.g. 404) are also cached but 5xx errors are not.

The value of this option can be any objects that implements method C<get> and
C<set> (e.g. C<CHI>), an existing directory for file caching, a true value to
enable global in-memory-caching, or a false value to disable caching (default).

File caching uses file names based on MD5 of an URL so for instance
C<http://example.org/> is cached as C<4389382917e51695b759543fdfd5f690.json>.

=item warn

Show error messages on the standard error.

=back

=head1 METHODS

=head2 time

Returns the UNIX timestamp right before the last request. This can be used for
instance to add timestamps or the measure how fast requests were responded.

=head2 construct_url( [ $base_url, ] $vars_url_or_path )

Returns an URL given a hash reference with variables, a plain URL or an URL
path. The optional first argument can be used to override option C<url>.

    $importer->construct_url( %query_vars )
    $importer->construct_url( $importer->url, %query_vars ) # equivalent

=head2 request($url)

Perform a HTTP GET request of a given URL including logging, caching, request
hook etc. Returns a hash/array reference or C<undef>.

=head1 EXTENDING

This importer provides two methods to filter requests and responses,
respectively. See L<Catmandu::Importer::Wikidata> for an example.

=head2 request_hook

Gets a whitespace-trimmed input line and is expected to return an unblessed
hash reference, an URL, or undef. Errors are catched and treated equal to
undef.

=head2 response_hook

Gets the queried response object and is expected to return an object.

=head1 LOGGING

URLs are emitted before each request on DEBUG log level.

=head1 LIMITATIONS

Future versions of this module may also support asynchronous HTTP fetching
modules such as L<HTTP::Async>, for retrieving multiple URLs at the same time.

=head1 SEE ALSO

L<Catmandu::Fix::get_json> provides this importer as fix function.

=encoding utf8

=head1 COPYRIGHT AND LICENSE

Copyright Jakob Voß, 2014-

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut
